### PR TITLE
Project 20: Authentication (Therese & Rebeka)

### PR DESCRIPTION
**Pull request by Thessan**
**Frontend:** [https://thessan-rebeka-auth-api-frontend.netlify.app](https://thessan-rebeka-auth-api-frontend.netlify.app)
**Backend:** [http://thessan-rebeka-auth-api.herokuapp.com](http://thessan-rebeka-auth-api.herokuapp.com)

**My reflections on how this week's project turned out:** Learnt a lot this week and it was super fun to work as a team. We tried out Material UI for the first time which made it a bit more challenging but feel we did very well. Happy that we made it so that Sign Up only has the function of signing up and then you have to Login to be able to see the userMessage that will be displayed on the authenticated page. 

What we still have to fix on our frontend however is that Signup- and login success/fail message still is attached when scrolling through the pages, they're not deleted.

### CODE REVIEW

File: frontend/src/components/UserMessage.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+        method: "GET",
+        headers: { Authorization: `${accessToken}` },
+    })
+        .then((response) => response.json())
+        .then((json) => {
+            setMessage(json);
+        });
+    return (
+        <>
+            <Message>{message}</Message>
```
Comment (JHarde) [inline frontend/src/components/UserMessage.js:25]: Nice and clear message when logging in! It only shows on the first login though, tried loggin in twice. Perhaps you guys can have a look at that :)

**Review by JHarde — APPROVED** [review]
We think you created a really nice project! It's really nice that you used material ui for this one. Great work in both the frontend and backend. It's very clear and user friendly. Great job!! 🏅 💯

File: backend/server.js [hunk 2]
```diff
@@ -2,28 +2,149 @@ import express from 'express'
...
+  if (!user.isModified('password')) {
+    return next();
+  }
+
+  //Hash the password, after the validation of sign up details
+  const salt = bcrypt.genSaltSync();
+  user.password = bcrypt.hashSync(user.password, salt);
+
+  next();
+})
```
Comment (KarolCislowski) [inline backend/server.js:49]: Great use of  `pre()`

File: frontend/src/components/LoginForm.js [hunk 3]
```diff
@@ -0,0 +1,132 @@
+import React, { useState } from 'react'
+import { useDispatch, useSelector } from 'react-redux'
+import { makeStyles } from '@material-ui/core/styles'
+import TextField from '@material-ui/core/TextField'
```
Comment (KarolCislowski) [inline frontend/src/components/LoginForm.js:4]: I love Material-UI as well ❤️

File: frontend/src/reducers/user.js [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+            state.login.statusMessage = statusMessage;
+        },
+        logout: (state, action) => {
+            state.login.isLoggedIn = false;
+            state.login.userId = 0;
+            state.login.accessToken = null;
+            state.login.statusMessage = "";
+        },
+    } 
+})
```
Comment (KarolCislowski) [inline frontend/src/reducers/user.js:35]: Reducer is a great place to manage your users but you need to store part of data in the local storage to make your users logged in after they refresh 🤓

**Review by KarolCislowski — CHANGES_REQUESTED** [review]
This is a great app 🦄 but you need to fix one little issue which will let your users logged in after refreshing the page.

Good luck 🌈

File: frontend/src/components/UserMessage.js (hunk 1 above)
Comment (Thessan) [reply frontend/src/components/UserMessage.js:25]: We can't see this error you're referring too. We've tried to login several times in a row and it works ☺️

**Comment by Thessan** [comment]
> This is a great app 🦄 but you need to fix one little issue which will let your users logged in after refreshing the page.
> 
> Good luck 🌈

**Comment by Thessan** [comment]
> This is a great app 🦄 but you need to fix one little issue which will let your users logged in after refreshing the page.
> 
> Good luck 🌈

Thanks! This has now been fixed! 😊

**Review by KarolCislowski — APPROVED** [review]

**Pull request closed without merging**